Different diffusion methods

Conni2461 opened the pull request.

hirschsn commented on repa/grids/diffusion.cpp:
```diff
@@ -391,5 +365,72 @@ void Diffusion::init_new_foreign_cell(local_cell_index_type localcell,
+        const std::string &impl = m[3].str();
+        if (comm_cart.rank() == 0)
+            std::cout << "Setting implementation to: " << impl << std::endl;
+        if (impl == "willebeek")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::WILLEBEEK);
+        else if (impl == "schornbaum")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::SCHORN);
+        else if (impl == "so")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::SO);
+        else if (impl == "sof")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::SOF);
```
Could you use a `std::map` or something like that for this? (Actually, I'd like to use a plain array and designated initializers, however, this feature is not available in C++14).

hirschsn replied:
Additional point (please consiver before implementing the map): Do we require that this instance is a PSDiffusion for any of these variants? If so, how do we enforce this?

Conni2461 replied:
I have just written a patch that addresses this problem, just need to test it. I'll do it in the morning and then commit if everything works.

Conni2461 replied:
Updated in 337159c.
I have changed it to use inheritance, as I believe that `Diffusion` does not need to know about `so` and `sof`. So I also moved the part with the beta setting to `PSDiffusion`. I also changed it so that the `Setting flow_count = 15` output is only generated when it really works. Also the errors are only displayed on rank 0, which will clean up the output.

Then I changed the macros with the dynamic_cast (DIFFUSION_MAYBE_SET_NFLOW_ITER and DIFFUSION_MAYBE_SET_BETA), because the try catch did not catch errors. This is because an incorrect cast returns a null pointer, which causes a segfault to be thrown because it tries to call a method on the null pointer. I don't know if other compilers would have caught it anyway, gcc-9.3 doesn't do it and with the new version it should be more robust.

hirschsn replied:
You're absolutely right, `dynamic_cast` only throws when casting references. For pointers the result is a nullptr if the cast fails.

hirschsn commented on repa/grids/psdiffusion.cpp:
```diff
@@ -0,0 +1,158 @@
+namespace grids {
+
+static Vec3i
+fix_periodic_edge(const Vec3i &c0, const Vec3i &c2, const Vec3i &gsize)
```
This function needs a better name. :)
"fix_periodic..." is its intention, but what does it do?

Conni2461 replied:
Updated in 14f7a25

hirschsn commented on repa/grids/psdiffusion.cpp:
```diff
@@ -0,0 +1,158 @@
+                = &neighborhood_ranks.data()[0] + (neighbors.size() * i);
+    }
+
+    // Neighborhood consistence check
```
consistency

Conni2461 replied:
Updated in 14f7a25

hirschsn commented on repa/grids/psdiffusion.cpp:
```diff
@@ -0,0 +1,158 @@
+    Diffusion::post_init(firstcall);
+    comm_dims = util::mpi_cart_get_dims(comm_cart);
+
+#ifdef PSDIFFUSION_DEBUG
```
Please, quickly comment here, that we save the neighborhood to check it later, don't we?

Conni2461 replied:
Updated in 14f7a25

hirschsn commented on repa/grids/psdiffusion.cpp:
```diff
@@ -0,0 +1,158 @@
+                                  rank_type neighrank) const
+{
+    const bool b1 = coords_based_allow_sending(cidx, neighrank);
+#ifdef PSDIFFUSION_DEBUG
```
For the merged version, please completely remove the `rank_based_allow_sending` function (as it is the slower one, right?).

Conni2461 replied:
Yes it is the slower version. Removed in 14f7a25

hirschsn commented on repa/grids/psdiffusion.hpp:
```diff
@@ -0,0 +1,57 @@
+namespace repa {
+namespace grids {
+
+struct PSDiffusion : public Diffusion {
```
Please add a comment to this class. (Same as diffusion but preserves structure, etc./what does that mean?)

Conni2461 replied:
Updated in 14f7a25

hirschsn commented on repa/grids/diffusion.cpp:
```diff
@@ -391,5 +365,72 @@ void Diffusion::init_new_foreign_cell(local_cell_index_type localcell,
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::SOF);
+        else
+            assert(false);
```
Better to print an error here that "`impl' is unknown".

Conni2461 replied:
Also fixed in the mentioned patch: https://github.com/hirschsn/repa/pull/36#discussion_r408405605

Conni2461 replied:
Updated in 337159c.

Currently an error is output and the program is terminated in the debug variant. 
In the release variant, the default variant (Willebeek) would be used here. We could consider whether we really want to do this, or whether we should end the program here as well, since the user has chosen a different implementation and may have made a typo.

hirschsn replied:
Yes, using the default implementation and issuing a warning is good.

Conni2461 commented on repa/grids/diffusion.cpp:
```diff
@@ -310,11 +278,17 @@ Diffusion::compute_send_list(std::vector<double> &&send_loads,
                                           std::end(neighbors), neighrank));
 
+            if (weights[cidx] <= 0
+                && send_loads[neighidx] < profit_percentage_pass_through * load)
```
That's a good point. I will run some simulations tomorrow and compare the current version with an unmodified copy of `send_loads`.
I will then add logs and results to the thread.

Conni2461 replied:
The 50k particles were simulated with a box size of 250^3 and 100 repart steps. Schornbaum with 15 iterations was used.
Summary of the [simulations](https://gist.github.com/Conni2461/7cd9b402df75eae16d097e72fae0d46b):

| Simulation                | Sum of repart time | Best imbalance | Levels out at |
|---------------------------|--------------------|----------------|---------------|
| default                   | 13.944s            | 1.780          | 1.839         |
| no_profit_check           | 6472.995s          | 1.012          | 1.012         |
| not_consistent_01_percent | 68.255s            | 1.064          | 1.064         |
| not_consistent_05_percent | 15.823s            | 1.780          | 1.839         |
| not_consistent_10_percent | 13.784s            | 1.772          | 1.839         |
| consistent_01_percent     | 60.691s            | 1.028          | 1.028         |
| consistent_05_percent     | 17.254s            | 1.780          | 1.839         |
| consistent_10_percent     | 14.574s            | 1.780          | 1.839         |

It seems like using an unaltered copy of `send_loads` is even better than the current version with less time and a better imbalance. If you also take a look at the logfiles you will see that the unaltered version of send_loads gets better faster.

hirschsn replied:
Thanks for the measurements! Quick help for me please: What exactly is "consistent" and "non_consistent" :)

Conni2461 replied:
consistent means unaltered copy of send_loads.
not_consistent means current commit.

hirschsn replied:
Okay, that means we use "consistent", i.e. an unaltered copy. Could you please commit that?

Conni2461 replied:
Done in 7e2e561.

hirschsn commented:
@Conni2461 I dropped the macros and changed them to inline functions that indicate if the corresponding request could be fulfilled or not.

hirschsn commented:
@Conni2461 Any ideas how we could include the variants in all tests? An option is to allow grid specifications "diff <variant>" and "<variant>" is then passed to command() directly after grid creation. We would, however, need to include the grid variants in the maps in grid_types, such that "supported_grid_types()" also returns them.

Conni2461 commented:
I will take a look into this later today.

Conni2461 commented:
I'm not sure about my result, so I decided not to commit it yet and release a [patch file](https://gist.github.com/Conni2461/13f30ae00ec661ef6e37553ebc6b8b68) instead.

First I added the `PSDiffusion` to the tests by adding the `grid_type` to the `unordered_map` in the file `repa/grid_types.cpp`. I had forgotten that before. After that the `tests/test_pos_rank_agreement.cpp` failed with `MPI_Abort`, so I adjusted that in the file `tests/test_pos_rank_agreement.cpp` by adding the `PS_DIFF` grid_type to the first and second run (I hope this is right).

I pulled the `static const unordered_map` of the supported variants for `Diffusion` and `PSDiffusion` into the classes and added the `virtual` method `get_supported_variants`. This returns a `set` of `strings`. For `Diffusion` this is `willebeek` and `schornbaum` and for `PSDiffusion` this is `willebeek`, `schornbaum`, `so` and `sof`.

Next, I moved everything in the method `TEnv::TEnv_impl::run` to a new method that takes place after the grid creation, so that I avoid code duplication. I cast the `unique_ptr` `up` with a `dynamic_cast` to `Diffusion` and if this is successful, I get the supported types, run `command("set flow" + impl)` on `up` (does not work with `d`, because `command` is protected in `Diffusion`) and then run the tests using the new method. If `up` is not of type `Diffusion`, I just run the new method.

Conni2461 commented:
I thought about it again and I think we have to recreate the `unique_ptr up` in the inner for loop(the loop which sets the flow variance with command) in every iteration. I wrote a simple [patch](https://gist.github.com/Conni2461/1b273809b5de3897c0ead60040f9b23e) for it, which must be applied after the other patch.

hirschsn commented:
@Conni2461 Both changes are fine. Could you please apply them both to this PR. I will, then, take another look at the changes.

hirschsn commented:
@Conni2461 I added a more general "variants" interface to make testenv independent of diffusion internals. Could you, please, quickly check if the changes are good?

hirschsn commented:
@Conni2461 for more than 8 processes, an assert triggers in debug builds in psdiffusion and more tests fail. Set TEST_MAX_NPROC to 16 to see which fail.

Conni2461 commented:
Your changes look good. But currently willebeek is executed twice, once before the for loop, as it is the default, and once in the loop. If it's all the same to you, it fits.

Conni2461 commented:
The tests fail because we currently have Cartesian1D set as default. I think it is because the grid is set too small. This change would fix the problem:
```diff
diff --git a/tests/testenv.cpp b/tests/testenv.cpp
index 762c331..e5e39fa 100644
--- a/tests/testenv.cpp
+++ b/tests/testenv.cpp
@@ -107,7 +107,7 @@ TEnv::TEnv_impl::TEnv_impl(repa::ExtraParams ep) : mings(1.0), ep(ep)
     repa::Vec3i dims{0, 0, 0};
     MPI_Dims_create(comm.size(), 3, dims.data());
     for (size_t i = 0; i < box.size(); ++i)
-        box[i] = mings * 5 * (dims[i] + 1);
+        box[i] = mings * 10 * (dims[i] + 1);
 }
 
 void TEnv::TEnv_impl::with_repart()
```
But we could also talk about setting the default grid partitioning to Cartesian3D. This would mean that we can no longer test for all processes 1-16, since PSDiffusion only works reliably if all have the same number of neighbors. I have run it with 16 processes and the following tests failed:
```
         19 - test_grid_types-13 (Failed)
         99 - test_neighborhood_symmetry-13 (Failed)
        163 - test_ghost_exchange_volume-13 (Failed)
        165 - test_ghost_exchange_volume-15 (Failed)
        191 - test_ghost_exchange_volume_ext-9 (Failed)
```
But this result is not even deterministic. 

A useful change is that when creating `dd` using `s.cell_system.set_new_dd` / `s.cell_system.set_generic_dd` you can specify the grid partition (linear, 1d, 3d). So you would not always have to recompile. If you think it makes sense I could take a look at this

hirschsn commented:
How are the number of neighbors different among the processes for a fully periodic grid?

Settings Cartesian3D is indeed a good default. Maybe we can make that the default and add a parameter in the constructor of GloMethod. Make_pargrid_impl could set this if it is supported by the constructor.

Conni2461 commented:
I was wrong. 3dcart uses 1d partitioning for 11 and 13 processes. The thing is, without enlarging the grid, we always have a process without neighbors. That means all processes say they have 2 neighbors and one has 0. We could put a check somewhere to make sure all processes have the same number of neighbors (1d, 3d). If this is not the case, we should definitely point out that the grid is not large enough for the given number of processes.

About the default 3d. I was thinking of something similar. I'll take a closer look at it later or tomorrow.

hirschsn commented:
Yes, 3dcart decomposes one-dimensionally if the number is prime. I still don't get where we would have zero neighbors in a 1d decomposition.

Conni2461 commented:
Applying
```diff
diff --git a/repa/grids/diffusion.cpp b/repa/grids/diffusion.cpp
index 31cb341..43101ad 100644
--- a/repa/grids/diffusion.cpp
+++ b/repa/grids/diffusion.cpp
@@ -227,7 +227,7 @@ Diffusion::Diffusion(const boost::mpi::communicator &comm,
     // Initial partitioning
     partition.resize(gbox.ncells());
     util::InitPartitioning{gbox, comm_cart}(
-        util::InitialPartitionType::CARTESIAN1D,
+        util::InitialPartitionType::CARTESIAN3D,
         [this](global_cell_index_type idx, rank_type r) {
             assert(r >= 0 && r < this->comm.size());
             this->partition[idx] = r;
@@ -371,6 +371,11 @@ void Diffusion::post_init(bool firstcall)
 {
     // Create graph comm with current process structure
     neighcomm = util::undirected_graph_communicator(comm_cart, neighbors);
+    if (firstcall) {
+        int nneigh = repa::util::mpi_undirected_neighbor_count(neighcomm);
+        std::cout << comm_cart.rank() << ": " << neighbors.size() << "/"
+                  << nneigh << std::endl;
+    }
 }
 
 void Diffusion::init_new_foreign_cell(local_cell_index_type localcell,
```
and running: `mpiexec -n 11 --oversubscribe ./tests/test_grid_types` will result in something like this:
```
Checking grid 'diff'
8: 2/2
3: 2/2
9: 2/2
5: 2/2
1: 2/2
0: 2/2
10: 0/0
2: 2/2
4: 2/2
6: 2/2
7: 2/2
```

For setting Cartesian1D i think we have a different problem, because here i get the same amount of neighbors for all processes.
I think if the grid is too small we end up with something like this:
![2020-04-23_10-41](https://user-images.githubusercontent.com/15233006/80089867-e1352d00-855e-11ea-8193-5cf9d7e4f3aa.png)
A carries out the check for the green cell and receives it. C executes the check for one of the red cells and gets one of the cells at the same time. But after that we have a new neighborhood between A and C and the assert fails when the next repart is called.

hirschsn commented:
@Conni2461 The problem with 0 neighbors is because in this case process 10 has no cells, so 0 neighbors is correct. I have a commit coming up that will distribute `CARTESIAN3D` more evenly.
We should, nevertheless, detect this case (0 local cells) in psdiffusion, issue an error and abort.

hirschsn commented:
Regarding the second issue: I see how this can happen. We have to directly mark the cell as owned by a different process in the `partition` array in `compute_send_list`.

Could you try to just drop the call to `mark_new_owners_from_sendvolume` in diffusion.cpp:181 and directly set `partition[i]` in `compute_send_list` (diffusion:cpp:314)?

Conni2461 commented:
Setting the `partition` array in `compute_send_list` doesn't change the tests and I still lose neighbors in 1D with more than 8 processes. Don't we still need to communicate the array in the neighborhood so that the problem is fixes.
My changes:
```diff
diff --git a/repa/grids/diffusion.cpp b/repa/grids/diffusion.cpp
index 31cb341..c83cfdf 100644
--- a/repa/grids/diffusion.cpp
+++ b/repa/grids/diffusion.cpp
@@ -177,9 +177,6 @@ bool Diffusion::sub_repartition(CellMetric m, CellCellMetric ccm)
     const auto send_information
         = std::make_pair(std::cref(cells_to_send), std::cref(neighbors));
 
-    // Update partition array
-    _impl::mark_new_owners_from_sendvolume(partition, send_information);
-
     //
     // First communication step
     // Send *all* vectors in "cells_to_send" to *all* neighbors.
@@ -251,7 +248,7 @@ void Diffusion::set_variant(const std::string &var)
 
 Diffusion::PerNeighbor<Diffusion::GlobalCellIndices>
 Diffusion::compute_send_list(std::vector<double> &&send_loads,
-                             const std::vector<double> &weights) const
+                             const std::vector<double> &weights)
 {
     std::vector<std::tuple<int, double, local_cell_index_type>> plist;
 
@@ -311,6 +308,7 @@ Diffusion::compute_send_list(std::vector<double> &&send_loads,
             if (weights[cidx] <= send_loads[neighidx]) {
                 to_send[neighidx].push_back(cells[cidx]);
                 send_loads[neighidx] -= weights[cidx];
+                partition[cells[cidx]] = neighrank;
                 // This cell is done. Continue with the next.
                 break;
             }
diff --git a/repa/grids/diffusion.hpp b/repa/grids/diffusion.hpp
index 09ca472..d89783f 100644
--- a/repa/grids/diffusion.hpp
+++ b/repa/grids/diffusion.hpp
@@ -127,7 +127,7 @@ private:
      */
     PerNeighbor<GlobalCellIndices>
     compute_send_list(std::vector<double> &&send_volume,
-                      const std::vector<double> &weights) const;
+                      const std::vector<double> &weights);
 
     /** Generate neighborhood information to send to processes that received
      * the cells in "cells_to_send".
```

hirschsn commented:
One fix would be to dup the communicator and in case of 1d initialization, replace it by a 1d cartesian communicator. Then, we can use this communicator in coord_based_allow_sending.

hirschsn commented:
Another note: Diffusion could use LinearInitPartitioning and then coord_bases_allow would not work at all. In this case, we either need to use the other transfer acceptance criterion or we need to prohibit PSDiffusion being initialized non-Cartesian.

hirschsn commented on repa/grids/psdiffusion.cpp:
```diff
@@ -0,0 +1,177 @@
+
+        cn = map_coords_to_opposite_side(c0, cn, comm_dims);
+        c2 = map_coords_to_opposite_side(c0, c2, comm_dims);
+        if (std::abs(cn[0] - c2[0]) > 2 || std::abs(cn[1] - c2[1]) > 2
```
Also, these `> 2` should be `>= 2`.

Conni2461 replied:
You are right. Good catch

hirschsn commented:
To clarify my posts. The following should work if Diffusion does a 1D Cartesian init:

```diff
diff --git a/repa/grids/diffusion.cpp b/repa/grids/diffusion.cpp
index 31cb341..c83cfdf 100644
--- a/repa/grids/diffusion.cpp
+++ b/repa/grids/diffusion.cpp
@@ -177,9 +177,6 @@ bool Diffusion::sub_repartition(CellMetric m, CellCellMetric ccm)
     const auto send_information
         = std::make_pair(std::cref(cells_to_send), std::cref(neighbors));
 
-    // Update partition array
-    _impl::mark_new_owners_from_sendvolume(partition, send_information);
-
     //
     // First communication step
     // Send *all* vectors in "cells_to_send" to *all* neighbors.
@@ -251,7 +248,7 @@ void Diffusion::set_variant(const std::string &var)
 
 Diffusion::PerNeighbor<Diffusion::GlobalCellIndices>
 Diffusion::compute_send_list(std::vector<double> &&send_loads,
-                             const std::vector<double> &weights) const
+                             const std::vector<double> &weights)
 {
     std::vector<std::tuple<int, double, local_cell_index_type>> plist;
 
@@ -311,6 +308,7 @@ Diffusion::compute_send_list(std::vector<double> &&send_loads,
             if (weights[cidx] <= send_loads[neighidx]) {
                 to_send[neighidx].push_back(cells[cidx]);
                 send_loads[neighidx] -= weights[cidx];
+                partition[cells[cidx]] = neighrank;
                 // This cell is done. Continue with the next.
                 break;
             }
diff --git a/repa/grids/diffusion.hpp b/repa/grids/diffusion.hpp
index 09ca472..d89783f 100644
--- a/repa/grids/diffusion.hpp
+++ b/repa/grids/diffusion.hpp
@@ -127,7 +127,7 @@ private:
      */
     PerNeighbor<GlobalCellIndices>
     compute_send_list(std::vector<double> &&send_volume,
-                      const std::vector<double> &weights) const;
+                      const std::vector<double> &weights);
 
     /** Generate neighborhood information to send to processes that received
      * the cells in "cells_to_send".
diff --git a/repa/grids/psdiffusion.cpp b/repa/grids/psdiffusion.cpp
index 8cd7743..d9b3c57 100644
--- a/repa/grids/psdiffusion.cpp
+++ b/repa/grids/psdiffusion.cpp
@@ -65,6 +65,13 @@ PSDiffusion::PSDiffusion(const boost::mpi::communicator &comm,
                          double min_cell_size)
     : Diffusion(comm, box_size, min_cell_size)
 {
+    Vec3i dims{0, 1, 1}, periods{1, 1, 1};
+    MPI_Dims_create(comm.size(), 3, dims.data());
+
+    MPI_Comm _comm_cart;
+    MPI_Cart_create(comm, 3, dims.data(), periods.data(), false, &_comm_cart);
+
+    comm_i = boost::mpi::communicator(_comm_cart, boost::mpi::comm_take_ownership);
 }
 
 PSDiffusion::~PSDiffusion()
@@ -120,20 +127,20 @@ bool PSDiffusion::coords_based_allow_sending(local_cell_index_type c,
                                              rank_type neighrank) const
 {
     Vec3i c0, cn;
-    MPI_Cart_coords(comm_cart, rank_of_cell(cells[c]), 3, c0.data());
-    MPI_Cart_coords(comm_cart, neighrank, 3, cn.data());
+    MPI_Cart_coords(comm_i, rank_of_cell(cells[c]), 3, c0.data());
+    MPI_Cart_coords(comm_i, neighrank, 3, cn.data());
     for (const global_cell_index_type &d :
          gbox.full_shell_neigh_without_center(cells[c])) {
         rank_type rank_d = rank_of_cell(d);
         if (rank_d == rank_of_cell(cells[c]) || rank_d == neighrank)
             continue;
         Vec3i c2;
-        MPI_Cart_coords(comm_cart, rank_d, 3, c2.data());
+        MPI_Cart_coords(comm_i, rank_d, 3, c2.data());
 
         cn = map_coords_to_opposite_side(c0, cn, comm_dims);
         c2 = map_coords_to_opposite_side(c0, c2, comm_dims);
-        if (std::abs(cn[0] - c2[0]) > 2 || std::abs(cn[1] - c2[1]) > 2
-            || std::abs(cn[2] - c2[2]) > 2)
+        if (std::abs(cn[0] - c2[0]) >= 2 || std::abs(cn[1] - c2[1]) >= 2
+            || std::abs(cn[2] - c2[2]) >= 2)
             return false;
     }
     return true;
diff --git a/repa/grids/psdiffusion.hpp b/repa/grids/psdiffusion.hpp
index 1c5b0de..1c226eb 100644
--- a/repa/grids/psdiffusion.hpp
+++ b/repa/grids/psdiffusion.hpp
@@ -53,7 +53,7 @@ private:
                                     rank_type neighrank) const;
 
     Vec3i comm_dims;
-
+    boost::mpi::communicator comm_i;
 #ifndef NDEBUG
     std::vector<rank_type> initial_neighborhood;
 #endif

```

Conni2461 commented:
Everything looks good and test are working fine. Will you commit the changes since you wrote them too?
Tomorrow I will finish the changes that allow to set an initial partition over the python script and then I think the PR is done.

hirschsn commented:
You go ahead and push the initial partition changes. I will, then, fix the accept-transfer comm bug.

Conni2461 commented:
The changes are committed. Please take a look at them. I also opened a pull-request at gitlab for the changes in espresso. I wasn't sure which branch, so I took the default(generic_dd) and adjusted it.

You should probably take a look at `graph.cpp`, because you could set the initial partition here. I adjusted it so that the functionality doesn't change, so before the changes it was `CART1D` and after the changes it is still `CART1D`. If you want to set it as well, you have to change the constructor and the `pargrid_factory` accordingly.

We could now make changes to the `testenv`, that we test all `InitialPartitionType`s here. We could also test the `UnknownPartitionTypeError` in `test_unknow_exceptions`. Probably we should also split the `initial_partitioning.hpp` into `hpp` and `cpp`, now that there is a `std::unordered_map` there.

hirschsn commented:
Could you move the unordered_map and the parsing functions to grid_types.[ch]pp?

And then, add a single test that tests different initial partitionings for one grid type.

Do we currently throw an error if PSDiffusion is used with "Linear"? If not, please add it.

Conni2461 commented:
Including `util/initial_partitioning.hpp' in `grid_types.hpp' (for the `enum`) causes espresso to stop building. The returned error is `_compat.hpp` is not found.
```
Scanning dependencies of target GenericDD
[  1%] Building CXX object src/core/generic-dd/CMakeFiles/GenericDD.dir/generic_dd.cpp.o
In file included from /home/conni/repos/uni/repa/repa/grids/util/initial_partitioning.hpp:22,
                 from /home/conni/repos/uni/repa/repa/grid_types.hpp:24,
                 from /home/conni/repos/uni/repa/repa/repa.hpp:22,
                 from /home/conni/repos/uni/espresso-add/src/core/generic-dd/generic_dd.cpp:27:
/home/conni/repos/uni/repa/repa/grids/util/../globox.hpp:27:10: fatal error: _compat.hpp: No such file or directory
   27 | #include "_compat.hpp"
      |          ^~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/core/generic-dd/CMakeFiles/GenericDD.dir/build.make:83: src/core/generic-dd/CMakeFiles/GenericDD.dir/generic_dd.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:4347: src/core/generic-dd/CMakeFiles/GenericDD.dir/all] Error 2
make: *** [Makefile:161: all] Error 2
```
 Changing the path of `_compat.hpp` causes espresso to not find other headers, etc. I could not explain and fix the error. So I made an `int-enum` from the `enum` and forward declared it.

`vec_arith.hpp` was missing a `pragma once` which caused me problems in the previous commit but I did not realize it in time. I hope the missing of `pragma once` is not intended.

hirschsn commented:
Okay, thanks. I will look into this PR and probably add some more (stylistic) changes as well as fix the coords-based-allow function.

Indeed, vec_arith.hpp is missing a pragma once. Initially, that was actually intentional. I did not want this header to be included in other headers. However, for now, adding it is good.

hirschsn commented:
Looks good to me. Any more remarks/additions/errors/... from you @Conni2461?

Conni2461 commented:
No further comments from my side. Seems fine to me.
